### PR TITLE
Update file paths

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Problem:
Paths for config file, logs, db, etc do not match the requirements [in the spec doc](https://docs.google.com/document/d/1MyT4Wt_-vgPwhwEjC-oeHRdLYrZWaQzFGTwclT7hD1U/edit?usp=sharing)

## Solution:
Update file paths and create the necessary parent directories

## Testing 
```
snapcraft clean
snapcraft pack
sudo snap install --devmode charmed-mongodb_5.0_amd64.snap
sudo snap start charmed-mongodb.mongod
charmed-mongodb.mongo
```

### CODE REVIEW

File: snap/hooks/install [hunk 1]
```diff
@@ -2,32 +2,33 @@
 
 set -eux
 
-# Create the config file that is used by pbm
-PBM_CONFIG_FILE="${SNAP_COMMON}/pbm_config.yaml"
+# Create the necessary parent directories 
+mkdir -p "${SNAP_DATA}/etc/pbm/"
+mkdir -p "${SNAP_DATA}/etc/mongod/"
+mkdir -p "${SNAP_COMMON}/var/lib/mongod/db"
```
Comment (delgod) [inline snap/hooks/install]:
```suggestion
mkdir -p "${SNAP_COMMON}/var/lib/mongod"
```

File: snap/hooks/install [hunk 2]
```diff
@@ -2,32 +2,33 @@
...
-# we need to create a child dir and give it access.
-mkdir -p "${SNAP_DATA}/pbm-agent-dir"
-
 # Copy over the mongod.conf and create needed directories/files
-MONGO_CONFIG_FILE="${SNAP_COMMON}/mongod.conf"
+MONGO_CONFIG_FILE="${SNAP_DATA}/etc/mongod/mongod.conf"
 echo "configuration file does not exist."
 echo "copying default config to ${MONGO_CONFIG_FILE}"
-cp -r ${SNAP}/etc/mongod.conf ${SNAP_COMMON}
+cp -r ${SNAP}/etc/mongod.conf ${MONGO_CONFIG_FILE}
```
Comment (delgod) [inline snap/hooks/install:19]: any reason why we copy? not move
```suggestion
mv ${SNAP}/etc/mongod.conf ${MONGO_CONFIG_FILE}
```

File: snap/hooks/install [hunk 3]
```diff
@@ -2,32 +2,33 @@
...
-sed -i "s:/var/log/mongodb/mongod.log:$SNAP_DATA/mongod.log:g" $SNAP_COMMON/mongod.conf
-sed -i "s:/var/lib/mongodb:$SNAP_COMMON/db:g" $SNAP_COMMON/mongod.conf
-sed -i "s:/var/run/mongod.pid:$SNAP_DATA/mongod.pid:g" $SNAP_COMMON/mongod.conf
+touch ${SNAP_COMMON}/var/log/mongod/mongod.log
+touch ${SNAP_COMMON}/var/lib/mongod/mongod.pid
+
+# mongod.conf default values are not consistent with the snap directory system.
+sed -i "s/fork: true/fork: false/g" $MONGO_CONFIG_FILE
+sed -i "s:/var/log/mongodb/mongod.log:$SNAP_COMMON/var/log/mongod/mongod.log:g" $MONGO_CONFIG_FILE
+sed -i "s:/var/lib/mongodb:$SNAP_COMMON/var/lib/mongod/db:g" $MONGO_CONFIG_FILE
```
Comment (delgod) [inline snap/hooks/install]:
```suggestion
sed -i "s:/var/lib/mongodb:$SNAP_COMMON/var/lib/mongod:g" $MONGO_CONFIG_FILE
```

File: snap/hooks/install [hunk 4]
```diff
@@ -2,32 +2,33 @@
...
-mkdir -p ${SNAP_COMMON}/db
-touch ${SNAP_DATA}/mongod.log
-touch ${SNAP_DATA}/mongod.pid
-
-
-sed -i "s/fork: true/fork: false/g" $SNAP_COMMON/mongod.conf
-sed -i "s:/var/log/mongodb/mongod.log:$SNAP_DATA/mongod.log:g" $SNAP_COMMON/mongod.conf
-sed -i "s:/var/lib/mongodb:$SNAP_COMMON/db:g" $SNAP_COMMON/mongod.conf
-sed -i "s:/var/run/mongod.pid:$SNAP_DATA/mongod.pid:g" $SNAP_COMMON/mongod.conf
+touch ${SNAP_COMMON}/var/log/mongod/mongod.log
```
Comment (delgod) [inline snap/hooks/install]: I believe we should give privileges for writing into `${SNAP_COMMON}/var/{lib,log}/<component>` dirs for 584788:root instead of pre-creation of files

File: snap/hooks/install [hunk 5]
```diff
@@ -2,32 +2,33 @@
...
+touch ${SNAP_COMMON}/var/lib/mongod/mongod.pid
+
+# mongod.conf default values are not consistent with the snap directory system.
+sed -i "s/fork: true/fork: false/g" $MONGO_CONFIG_FILE
+sed -i "s:/var/log/mongodb/mongod.log:$SNAP_COMMON/var/log/mongod/mongod.log:g" $MONGO_CONFIG_FILE
+sed -i "s:/var/lib/mongodb:$SNAP_COMMON/var/lib/mongod/db:g" $MONGO_CONFIG_FILE
+sed -i "s:/var/run/mongod.pid:$SNAP_COMMON/var/lib/mongod/mongod.pid:g" $MONGO_CONFIG_FILE
```
Comment (delgod) [inline snap/hooks/install]: Is there any reason why pid file should be preserved?

File: snap/hooks/install [hunk 6]
```diff
@@ -2,32 +2,33 @@
...
-touch ${SNAP_DATA}/mongod.log
-touch ${SNAP_DATA}/mongod.pid
-
-
-sed -i "s/fork: true/fork: false/g" $SNAP_COMMON/mongod.conf
-sed -i "s:/var/log/mongodb/mongod.log:$SNAP_DATA/mongod.log:g" $SNAP_COMMON/mongod.conf
-sed -i "s:/var/lib/mongodb:$SNAP_COMMON/db:g" $SNAP_COMMON/mongod.conf
-sed -i "s:/var/run/mongod.pid:$SNAP_DATA/mongod.pid:g" $SNAP_COMMON/mongod.conf
+touch ${SNAP_COMMON}/var/log/mongod/mongod.log
+touch ${SNAP_COMMON}/var/lib/mongod/mongod.pid
```
Comment (delgod) [inline snap/hooks/install]: is there any reason why pid file should be preserved in common dir?

**Review by delgod — CHANGES_REQUESTED** [review]

File: snap/hooks/install (hunk 2 above)
Comment (MiaAltieri) [reply snap/hooks/install:19]: Good question, I am sure moving would be fine instead of copying. @jardon is that okay?

File: snap/hooks/install (hunk 5 above)
Comment (MiaAltieri) [reply snap/hooks/install]: I *think* it is actually not necessary - seems like we keep `processManagement.fork` [as its default value `false`](https://www.mongodb.com/docs/manual/reference/configuration-options/) and if we keep it at it's default we don't need to specify the `mongod.pid` file. But I could be wrong, mind sharing your knowledge here @jardon ?

File: snap/hooks/install (hunk 6 above)
Comment (MiaAltieri) [reply snap/hooks/install]: I wasn't sure about this, but my reasoning was that according [to the spec document](https://docs.google.com/document/d/1MyT4Wt_-vgPwhwEjC-oeHRdLYrZWaQzFGTwclT7hD1U/edit?usp=sharing):
> data - the core workload data for the service running
For Kafka, this will be the topics, metadata, offsets etc
For ZooKeeper, this will be the state snapshots
The snap directory for this should $SNAP_COMMON/var/lib/COMPONENT/

I figured the `pid` was `metadata` that was relevant to the workload, and hence should go in `$SNAP_COMMON/var/lib/COMPONENT/`. But I wasn't entirely sure 😅  is there a more appropriate place?

Comment (MiaAltieri) [reply snap/hooks/install]: Ah I misunderstood the docs, it is better in conf files i.e. `$SNAP_DATA/etc/COMPONENT`- but lets figure out if we need this first

File: snap/hooks/install (hunk 2 above)
Comment (jardon) [reply snap/hooks/install:19]: What's the default behavior of `mv` when the source location is a read-only filesystem?  Copying the file ensures that you have a backup of a default config in case you need it for any reason.  This way you wouldnt have to write it from scratch later and have to figure out what defaults work for the snap environment.

File: snap/hooks/install (hunk 5 above)
Comment (jardon) [reply snap/hooks/install]: This was required to configure `mongod` to not create the pid file under `/var/run` as that is not possible with snaps.

**Review by jardon — APPROVED** [review]

File: snap/hooks/install (hunk 5 above)
Comment (delgod) [reply snap/hooks/install]: is `/var/lib` is read-only, please move pid file to `/tmp`

File: snap/snapcraft.yaml [hunk 7]
```diff
@@ -16,17 +16,17 @@ system-usernames:
 
 layout:
   /etc/mongod.conf:
-    symlink: $SNAP_COMMON/mongod.conf
+    symlink: $SNAP_DATA/etc/mongod/mongod.conf
```
Comment (delgod) [inline snap/snapcraft.yaml]: Why is that line needed? I am not sure

**Review by delgod — CHANGES_REQUESTED** [review]

File: snap/hooks/install (hunk 5 above)
Comment (jardon) [reply snap/hooks/install]: That doesn't work. Its moving it to `$SNAP_COMMON/var/lib` which is writable by the snap.  `/tmp` doesnt make sense in this context by itself.  Maybe `$SNAP_COMMON/tmp`, but otherwise it would need a `system-files` interface to do that.

File: snap/snapcraft.yaml (hunk 7 above)
Comment (jardon) [reply snap/snapcraft.yaml]: This likely isnt needed.  I am fairly certain that the layout here is not utilized and can be removed.  This would need to be tested though.

Comment (MiaAltieri) [reply snap/snapcraft.yaml]: Good to know, testing now :)

**Review by delgod — APPROVED** [review]
